iceoryx_utils/include/iceoryx_utils/cxx/string.hpp

```diff
@@ -284,7 +373,7 @@ class string
     /// @todo consider implementing the inequality operator for a char array for which the size is known at compile
     /// time; it could have the following signature
     /// template <int N>
-    /// bool operator==(const char (&rhs)[N]) const noexcept
+    /// bool operator!=(const char (&rhs)[N]) const noexcept
```

Comment:
This we actually could handle. A string literal of size `N` needs to have a `\0` terminator at the end. Therefore we could check if the string literal is valid like this:
```
template <int N>
bool isValidStringLiteral(const char (&rhs)[N]) {
    return rhs[N] == `\0\; // this would produce a warning since we are accessing a element out of bounds but in our special case it would be valid since string literals have always an additional element, the terminator
}
```

If you then would have the following evil misuse case:
```
char bla[3] = "bl";
bla[3] = 'a';
isValidStringLiteral(bla); // -> returns false
```

@elBoberido should we integrate something like this into the string or is it only something Machiavelly would do?

Reply:
we have to differentiate here, `"bl"` is a string literal with the type of `const char[3]` and is always null terminated.
On the other side, `char bla[3]` is just a C-style string which can be mutated and Murphy can mess up and overwrite the null termination. There was also #264 which was fixed last week. I'm not fully happy with that fix, since it could also have been fixed by an overload of `char[N]`. But then many tests would have failed and since a non null terminated C-style string is invalid anyhow, I took that approach. Have to discuss it with @FerdinandSpitzschnueffler, though.

Reply:
Fine with me to handle this later. I would like to propose a stringref anyway and there we also need to take a look at operations like this.

And I was referring to the issue @elBoberido  described, not string literals.